### ISSUE TITLE
Cannot filter out the APIs 

### ISSUE THREAD

**Issue by Kavignon** [dotnet/docfx#3665]
**DocFX Version Used**: 
2.40.2

**Template used**: (`default` or `statictoc` or contain custom template)
statistic

**Steps to Reproduce**:

1. Created a filter document (filterConfig.yml)
2. Included APIs
apiRules:

-include: 
    uidRegex: ^Kevin\.Avignon\.People\.Namespace\.Person
    type: Class
-include:
    uidRegex: ^Kevin\.Avignon\.People\.Namespace\.PeopleController\.GetPerson
    type: Method 

-exclude: 
     uidRegex: ^Kevin\.Avignon\.People
     type: Type

3. 

**Expected Behavior**:

I expect to see to the public documentation of Person class, the public documentation for GetPerson method and everything under the People namespace to be excluded.
**Actual Behavior**:
I only see the public documentation of Person.

This is a business case where I have changed the documentation content so I don't divulge any sensible information, but the gist is here.

**Comment by Kavignon** [comment]
apiRules:
- include: 
    uidRegex: ^Kevin\.Avignon\.People\.Controllers\.PeopleController\.GetPersonName.*$
    type: Method
- include: 
    uidRegex: ^Kevin\.Avignon\.People\.Controllers\.PeopleController\.HairColor.*$
    type: Property
- exclude: 
    uidRegex: ^Kevin\.Avignon\.People
    type: Type

Sorry there were some mistakes in the previous message I sent I'm sorry for this @superyyrrzz 

I did consult this page and I've been consulting it while creating my filter file. 


My use case is to only document what is found in the filter file. 

I've debug the code of docFX and it would seem that if my apiRule doesn't include a class, but points to an member within that said class (i.e. Method, Property for instance), it doesn't get documented by docFX.

Most of my solution does not have to be document which is why by the end I follow what's mentioned in your documentation: 

> you need to put the most detailed rule in the top. 

Which is why I make my specific inclusion/exclusion on top of my filterConfig.yml and at the end, we can find that anything else that isn't worth mentioning should **not** be documented. 

That's the behaviour I expected from reading the documentation about filtering unwanted APIs.

**Comment by superyyrrzz** [comment]
@Kavignon Thank you for the feedback. Looks like we need to improve this document.

**Comment by Kavignon** [comment]
@superyyrrzz 

Something also that I'm still uneasy about it's concerning documentation of the internals of a container such as a 

- Type
- Class 
- Interface

When I ignore all found Type within a namespace except the ones mentioned in my inclusion, I do not expect what's within the container to be ignored. I'll make that statement clearer. 

```
- exclude: 
    uidRegex: ^Kevin\.Avignon\.People
    type: Type
```

Here, I ignore all classes and interfaces found under the 'People' namespace. 

Let's say there's a method, '_GetUndergradStudent_', found under the class _Teacher_ that can be retrieved at 

`Kevin.Avignon.People.School` (Bad namespace here, bear with me, I'm using this sample to make a point)

It can happen that a portion of the codebase is sensible and shouldn't be exposed. So, it makes sense, in my use case, to select carefully what needs to documented in our APIs. Debugging DocFX has shown me the following behavior: Before I can see a method under a class, I need to include said that class. That's not the behavior expected based on the _uidRegex_ associated with _type_. Whenever one of the high level containers is ignored, DocFX should be able to look under the definition of said container and look whether or not there could be something in there that **has** to be documented. 

Such as looking at the API rules and see if at least, one has a partial match with the signature of the class, and if so, then 'ignore' the ignore and take a look into the class. Since I have this API ruling that mentions to ignore them, I cannot document _GetUndergradStudent_ because first, we need to find _Teacher_ which will then lead to take a look within Teacher.

**Comment by superyyrrzz** [comment]
@Kavignon I agree with your concern. I think the existing behavior makes sense, while it does make the config error prone. To improve this:

1. We still need to improve the document, although it has been documented: 
    > If a namespace is excluded, all types/members defined in the namespace would also be excluded. If a type is excluded, all members defined in the type would also be excluded.

2. Better warning to hint users that some apiFilter rules are never hit.

**Comment by Kavignon** [comment]
@superyyrrzz I understand that, but also keep in mind that the documentation also mentions to put the most specific rules at the top. So, I should be able to get the members of my type even though it's ignored at the bottom,

I'm working on my fork at the moment with something such as `TryDocumentIncludedApiMember`
This method will be able to overcome this difficulty. Could that interest you as a pull request?

For sure, I'd also link it with a second pull request that would update the documentation for filtering APIs. I think that the previous way could hinder work for future other users of the tool and this feels more natural.

**Comment by Kavignon** [comment]
That way, every API rules gets hit and more to the point, the documentation can be changed for the following: 

> If a type is excluded, all members defined in the type would also be excluded unless specially specified as a specific rule at the top.

**Comment by superyyrrzz** [comment]
@Kavignon Thank you. Of course the pull request is welcomed. I agree this change can make the config easier. However, I still have some concerns about it:
1. Will it lead to obvious performance downgrade as it will need to go through each member, even the parent type is excluced? 
2. The default template assumes the YAML files are organized with namespace-class-member hierachy. I am afraid the template will not work properly if the parent is missing. It may also need update.

Anyway, we can look into it if the PR is ready.

**Comment by yufeih** [comment]
The cause of confusion is that when we are traversing the type hierarchy, 
if a higher level constructs are excluded, lower level constructs underneath it didn't get a chance to
be visited, even if they are explicitly marked as include.

In your example, even if `PeopleController.GetPerson` method is explicitly included, but because `PeopleController` is excluded,
`PeopleController.GetPerson` didn't get a chance to be processed at all.

This is indeed not a user intended behavior and I understand your pain. A workaround requires including the `PeopleController` type, and excluding all but `PeopleController.GetPerson` method underneath `PeopleController`.

I feel like the right code fix is not about the order of exclude, include rules, it is about removing the aforementioned restriction.

**Comment by Kavignon** [comment]
@yufei can you clarify what would mean with an example because I'm not entirely sure to understand what you meant in your previous comment?

**Comment by superyyrrzz** [comment]
@Kavignon It means to change from including what you want into exluding what you don't want, like:
```
apiRules:
- exclude: 
    uidRegex: ^Kevin\.Avignon\.People\.Controllers\.((?!PeopleController\.GetPersonName).)*$
    type: Method
- exclude: 
    uidRegex: ^Kevin\.Avignon\.People\.Controllers\.((?!PeopleController\.HairColor).)*$
    type: Property
```
Note that I do not explicitly include the type as it is already included by default.

Back to your PR, personally I do not agree with the following goal.
> If no rule is matched, the API will be excluded by default when a filter document is provided in tool's configuration

As the source code already express the intention whether to document some API through the accessor `private/public`. The config rule could be a patch on existing ones, but should not negelect and overwrite them. If user provides an empty config, I think the behavior should remain same with no rule config provided. 

In addition, this is a breaking change that we do not want to happen in v2. I am afraid we cannot accept this PR.

**Comment by Kavignon** [comment]
What would prefer then as a behavior if we keep in mind all that we've discussed for the past few months?

**Comment by superyyrrzz** [comment]
@Kavignon As discussed, I like your idea to make include rule effective, despite the exclude rule below it. I believe this is more intuitive behavior. I will be more happy if the behavior change is only limited to this.

However, regarding the second idea "If no rule is matched, the API will be excluded by default when a filter document is provided in tool's configuration", I'm not sure whether this is as intuitive as the first one for others. Moreover, this is a **breaking change** that I want to avoid. This is the baseline. Consider an existing API filter user finds it not work after upgrading DocFX, which is really frustrating.

**Comment by Kavignon** [comment]
So if I understand correctly, you would prefer my implementation when I can specificallly target members of a type? @superyyrrzz

**Comment by superyyrrzz** [comment]
@Kavignon Sure. I'd prefer a fix to address this issue instead of re-design the entire behavior. Our performance concern is also based on this goal.

**Comment by Kavignon** [comment]
@superyyrrzz Sorry I've been away from this for so long. I will go back on the details of the thread and see what I can do to make this a successful PR.
